contrib/etcd/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package adapter
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/cloudevents/sdk-go/pkg/cloudevents"
+	eventsclient "github.com/cloudevents/sdk-go/pkg/cloudevents/client"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/types"
+	"github.com/knative/eventing-sources/pkg/kncloudevents"
+	"github.com/knative/pkg/logging"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerAgentName = "resource-controller-adapter"
+	updateEventType     = "dev.knative.etcd.object.update"
+	deleteEventType     = "dev.knative.etcd.object.delete"
+
+	envSinkURI    = "SINK_URI"
+	envAPIVersion = "APIVERSION_"
+	envKind       = "KIND_"
+)
+
+// Add creates new EtcdSource controllers, one per resource,
+// and adds them to the manager.
+func Add(mgr manager.Manager) error {
+	sinkURI, defined := os.LookupEnv(envSinkURI)
+	if !defined {
+		return fmt.Errorf("required environment variable not defined '%s'", envSinkURI)
+	}
+
+	eventsClient, err := kncloudevents.NewDefaultClient(sinkURI)
+	if err != nil {
+		return err
+	}
+
+	// Create one controller per watch
+	i := 0
+	for {
```

Comment:
This is clever and maybe you would like to contribute it to https://github.com/kelseyhightower/envconfig